docs/decisions/0008-idempotency-keys.rst

```diff
@@ -0,0 +1,124 @@
+obvious uniqueness requirements that are easy to prove will apply to every case
+in the past, present and future.  However, it may not always be possible to
+construct a uniqueness constraint that is so complex that it adequately
+encapsulates all edge cases of the desired business logic.
+
+Decision
+********
+
+For all REST/python APIs to create Subsidy/Ledger, Transaction, and Reversal
+objects, incorporate idempotency keys to de-duplicate requests from the client.
+This should be implemented via an ``idempotency_key`` request body parameter.
+It should be stored as a like-named database field on all target objects, and
+there should be a unique constraint on that new db column.
+
+The client should generate idempotency keys for each object type as follows:
+
++--------------------+------------------+--------------------------------------------------------------------+---------------------------------+
+| Client             | Object to create | Idempotency key format                                             | Notes                           |
++====================+==================+====================================================================+=================================+
+| enterprise-subsidy | Ledger           | ``ledger-for-subsidy-{subsidy_uuid}``                              |                                 |
++--------------------+------------------+--------------------------------------------------------------------+---------------------------------+
+| enterprise-subsidy |                  | ``ledger-for-subsidy-{subsidy_uuid}-{quantity}-initial-deposit``   | First ledger credit only        |
++--------------------+ Transaction      +--------------------------------------------------------------------+---------------------------------+
+| enterprise-access  |                  | ``ledger-for-subsidy-{subsidy_uuid}-{hashed transaction data}`` Δ  | All subsequent ledger debits    |
++--------------------+------------------+--------------------------------------------------------------------+---------------------------------+
+|                    |                  | ``admin-invoked-reverse-{transaction_uuid}``                       | Reversals initiated by admins   |
+| enterprise-subsidy | Reversal         +--------------------------------------------------------------------+---------------------------------+
+|                    |                  | ``unenrollment-reversal-{fulfillment_identifier}-{unenrolled_at}`` | Reversals initiated by learners |
++--------------------+------------------+--------------------------------------------------------------------+---------------------------------+
+
+Δ = The hashed transaction metadata should incorporate the following inputs:
+* lms_user_id to enroll into a course.
+* content_key to enroll the learner into.
+* subsidy_access_policy_uuid pertaining to the policy which "granted" this redemption.
+* A list of all historical inactive redemptions for the learner and course.
```

Comment:
nit: I think you need a newline at the beginning of this block for it to render properly.